doc: update What is ACRN introduction (#6579)
projectacrn/acrn-hypervisor: doc/introduction/index.rst

pre-launched Safety VM.

Shown on the right of :numref:`V2-hl-arch`, the remaining hardware resources are
shared among the Service VM and User VMs. The Service VM is launched by the
hypervisor after any pre-launched VMs are launched. The Service VM can access
remaining hardware resources directly by running native drivers and provides
device sharing services to the User VMs, through the Device Model.  These
post-launched User VMs can run one of many OSs including Ubuntu, Android,
Windows, or a real-time OS such as Zephyr, VxWorks, or Xenomai. Because of its
real-time capability, a real-time VM (RTVM) can be used for software
programmable logic controller (PLC), inter-process communication (IPC), or
Robotics applications.  These shared User VMs could be impacted by a failure in
the Service VM since they may rely on its mediation services for device access.

The Service VM owns most of the devices including the platform devices, and
provides I/O mediation. The notable exceptions are the devices assigned to the
pre-launched User VM. Some PCIe devices may be passed through to the
post-launched User OSes via the VM configuration.

> The ACRN hypervisor also runs the ACRN VM manager to collect running

Author: no

information of the User VMs, and controls the User VMs such as starting,
stopping, and pausing a VM, and pausing or resuming a virtual CPU.

See the :ref:`hld-overview` developer reference material for more in-depth
information.

ACRN Device Model Architecture
******************************

Because devices may need to be shared between VMs, device emulation is
used to give VM applications (and their OSs) access to these shared devices.
Traditionally there are three architectural approaches to device

> emulation:

Author: I added some more material to cover this.

* **Device emulation within the hypervisor**: a common method implemented within
  the VMware\* workstation product (an operating system-based hypervisor). In
  this method, the hypervisor includes emulations of common devices that the
  various guest operating systems can share, including virtual disks, virtual
  network adapters, and other necessary platform elements.

* **User space device emulation**: rather than the device emulation embedded
  within the hypervisor, it is implemented in a separate user space application.
  QEMU, for example, provides this kind of device emulation also used by other
  hypervisors. This model is advantageous, because the device emulation is
  independent of the hypervisor and can therefore be shared for other
  hypervisors. It also permits arbitrary device emulation without having to
  burden the hypervisor (which operates in a privileged state) with this
  functionality.

> * **paravirtualized (PV) drivers**: a hypervisor-based device emulation model

Reviewer: ```suggestion
* **Paravirtualized (PV) drivers**: a hypervisor-based device emulation model
```

  introduced by the `XEN Project`_. In this model, the hypervisor includes the
  physical device drivers, and each guest operating system includes a
  hypervisor-aware driver that works in concert with the hypervisor drivers.

.. _XEN Project:
   https://wiki.xenproject.org/wiki/Understanding_the_Virtualization_Spectrum

There's a price to pay for sharing devices. Whether device emulation is
performed in the hypervisor, or in user space within an independent VM, overhead
exists.  This overhead is worthwhile as long as the devices need to be shared by
multiple guest operating systems. If sharing is not necessary, then there are

> more efficient methods for accessing devices, for example "passthrough".

Reviewer 2: ```suggestion
more efficient methods for accessing devices, for example, "passthrough."
```

The ACRN device model is a placeholder of the User VM. It allocates memory for
the User OS, configures and initializes the devices used by the User VM, loads
the virtual firmware, initializes the virtual CPU state, and invokes the ACRN
hypervisor service to execute the guest instructions.  ACRN Device model is an
application running in the Service VM that emulates devices based on command
line configuration

See the :ref:`hld-devicemodel` developer reference for more information.

> Device Passthrough

Reviewer: We use the ter `guest` a lot in this paragraph. Shouldn't we use `User VM` or `VM` instead?

of a device to a given guest operating system so that the device can be
used exclusively by that guest.

.. figure:: images/device-passthrough.png
   :align: center
   :name: device-passthrough

   Device Passthrough

Near-native performance can be achieved by using device passthrough.  This is
ideal for networking applications (or those with high disk I/O needs) that have
not adopted virtualization because of contention and performance degradation
through the hypervisor (using a driver in the hypervisor or through the
hypervisor to a user space emulation).  Assigning devices to specific guests is
also useful when those devices inherently wouldn't be shared. For example, if a
system includes multiple video adapters, those adapters could be passed through
to unique guest domains.

Finally, there may be specialized PCI devices that only one guest domain uses,
so they should be passed through to the guest. Individual USB ports could be
isolated to a given domain too, or a serial port (which is itself not shareable)
could be isolated to a particular guest. In ACRN hypervisor, we support USB
controller passthrough only, and we don't support passthrough for a legacy

> serial port, (for example ``0x3f8``).

Reviewer 2: ```suggestion
serial port (for example, ``0x3f8``).
```

=========================================

By using the latest virtualization-enhanced processor architectures, hypervisors
and virtualization solutions can support device passthrough (using VT-d),
including Xen, KVM, and ACRN hypervisor.  In most cases, the guest operating
system in the User VM, must be compiled to support passthrough by using kernel
build-time options.

Static Configuration Based on Scenarios
***************************************

Scenarios are a way to describe the system configuration settings of the ACRN
hypervisor, VMs, and resources they have access to that meet your specific
application's needs such as compute, memory, storage, graphics, networking, and
other devices.  Scenario configurations are stored in an XML format file and
edited using the ACRN configurator tool.

Following a general embedded-system programming model, the ACRN hypervisor is
designed to be statically customized at build time per hardware and scenario,
rather than providing one binary for all scenarios.  Dynamic configuration
parsing is not used in the ACRN hypervisor for these reasons:

* **Reduce complexity**. ACRN is a lightweight reference hypervisor, built for

>   embedded IoT. As new platforms for embedded systems are rapidly introduced,

Reviewer: ```suggestion
  embedded IoT and Edge. As new platforms for embedded systems are rapidly introduced,
```

  support for one binary could require more and more complexity in the
  hypervisor, which is something we strive to avoid.
* **Maintain small footprint**. Implementing dynamic parsing introduces hundreds or
  thousands of lines of code. Avoiding dynamic parsing helps keep the
  hypervisor's Lines of Code (LOC) in a desirable range (less than 40K).
* **Improve boot time**. Dynamic parsing at runtime increases the boot time. Using a
  static build-time configuration and not dynamic parsing helps improve the boot
  time of the hypervisor.

> The scenario XML file together with a target HW board XML file are used to build

Reviewer 2: ```suggestion
The scenario XML file together with a target board XML file are used to build
```

> the ACRN hypervisor image tailored to your HW and application needs. The ACRN

Reviewer 2: ```suggestion
the ACRN hypervisor image tailored to your hardware and application needs. The ACRN
```

project provides a board inspector tool to automatically create the board XML
file by inspecting the target hardware. ACRN also provides a

> :ref:`configurator tool <acrn_configuration_tool>`

Reviewer: ```suggestion
:ref:`configuration tool <acrn_configuration_tool>`
```

Author: we're settled in on calling it by the tool's name (configurator) rather than its purpose (configuration)

to create and edit a tailored scenario XML file based on pre-defined sample
scenario configurations.

.. _usage-scenarios:

> Pre-Defined Sample Scenarios

Reviewer 2: Standardize on "pre-defined" or "predefined"?

****************************

Project ACRN provides some pre-defined sample scenarios to illustrate how you
can define your own configuration scenarios.

> * **Industry** is a traditional computing, memory, and device resource sharing

Reviewer: ```suggestion
* **Shared** is a traditional computing, memory, and device resource sharing
```

Author: Alas, it's still called "Industry" in 2.6 but will be called "shared" in 2.7

  model among VMs. The ACRN hypervisor launches the Service VM. The Service VM
  then launches any post-launched User VMs and provides device and resource
  sharing mediation through the Device Model.  The Service VM runs the native
  device drivers to access the hardware and provides I/O mediation to the User
  VMs.

  .. figure:: images/ACRN-industry-example.png
     :width: 700px
     :align: center
     :name: arch-shared-example

>      ACRN High-Level Architecture Industry (Shared) Example

Reviewer: ```suggestion
     ACRN High-Level Architecture Shared Example
```

* **Partitioned** is a VM resource partitioning model when a User VM requires
  independence and isolation from other VMs.  A partitioned VM's resources are
  statically configured and are not shared with other VMs.  Partitioned User VMs

>   can be Real-time VMs, Safety VMs, or standard VMs and are launched at boot

Reviewer: ```suggestion
  can be Real-Time VMs, Safety VMs, or standard VMs and are launched at boot
```

  time by the hypervisor. There is no need for the Service VM or Device Model
  since all partitioned VMs run native device drivers and directly access their
  configured resources.

  .. figure:: images/ACRN-partitioned-example.png
     :width: 700px
     :align: center
     :name: arch-partitioned-example

     ACRN High-Level Architecture Partitioned Example

* **Hybrid** scenario simultaneously supports both sharing and partitioning on
  the consolidated system. The pre-launched (partitioned) User VMs, with their
  statically configured and unshared resources, are started by the hypervisor.
  The hypervisor then launches the Service VM. The post-launched (shared) User
  VMs are started by the Device Model in the Service VM and share the remaining
  resources.

  .. figure:: images/ACRN-hybrid-rt-example.png
     :width: 700px
     :align: center
     :name: arch-hybrid-rt-example

     ACRN High-Level Architecture Hybrid-RT Example

You can find the pre-defined scenario XML files in the
:acrn_file:`misc/config_tools/data` folder in the hypervisor source code. The
:ref:`acrn_configuration_tool` tutorial explains how to use the ACRN

> configurator to create your own scenario, or to view and modify an existing one.

Reviewer: ```suggestion
configuration tool to create your own scenario, or to view and modify an existing one.
```

   the ACRN Device Model and Virtual bootloader through ``dm-verity``.
#. The virtual bootloader starts the User-side verified boot process.

In this boot mode, the boot options of a pre-launched VM and the Service VM are defined
in the variable of ``bootargs`` of struct ``vm_configs[vm id].os_config``
in the source code ``configs/scenarios/$(SCENARIO)/vm_configurations.c`` (which
resides under the hypervisor build directory) by default.
These boot options can be overridden by the GRUB menu. See :ref:`using_grub` for
details. The boot options of a post-launched VM are not covered by hypervisor

> source code or a GRUB menu; they are defined in a guest image file or specified by

Reviewer 2: Although we are not changing "guest" widely for the 2.6 release, I suggest changing it in this one case for clarity. Suggestion:
```suggestion
source code or a GRUB menu; they are defined in the VM's OS image file or specified by
```

launch scripts.

`Slim Bootloader`_ is an alternative boot firmware that can be used to
boot ACRN. The `Boot ACRN Hypervisor
<https://slimbootloader.github.io/how-tos/boot-acrn.html>`_ tutorial
provides more information on how to use SBL with ACRN.

Learn More
**********

The ACRN documentation offers more details of topics found in this introduction
about the ACRN hypervisor architecture, Device Model, Service VM, and more.

These documents provide introductory information about development with ACRN:

> * :ref:`introduction`

Reviewer 2: Remove this line, the file is pointing to itself.

Author: doh
